pallets/loan/src/lib.rs

```diff
@@ -819,6 +977,97 @@ impl<T: Config> Pallet<T> {
 		);
 		Ok(nav)
 	}
+
+	/// writes off a given unhealthy loan
+	/// if override_write_off_index is Some, this is a admin action and loan override flag is set
+	/// if loan is already overridden and override_write_off_index is None, we return error
+	/// if loan is still healthy, we return an error
+	/// loan is accrued and nav is updated accordingly
+	/// returns new write off index applied to loan
+	fn write_off(
+		pool_id: T::PoolId,
+		loan_id: T::LoanId,
+		override_write_off_index: Option<u32>,
+	) -> Result<u32, DispatchError> {
+		LoanInfo::<T>::try_mutate(
+			pool_id,
+			loan_id,
+			|maybe_loan_data| -> Result<u32, DispatchError> {
+				let mut loan_data = maybe_loan_data.take().ok_or(Error::<T>::ErrMissingLoan)?;
+				// ensure loan is active
+				ensure!(
+					loan_data.status == LoanStatus::Active,
+					Error::<T>::ErrLoanNotActive
+				);
+
+				let now = Self::time_now()?;
+
+				// ensure loan was not overwritten by admin and try to fetch a valid write off group for loan
+				let write_off_groups = PoolWriteOffGroups::<T>::get(pool_id);
+				let write_off_group_index =
+					match (loan_data.admin_written_off, override_write_off_index) {
+						// admin is trying to write off
+						(_admin_written_off, Some(index)) => {
+							// check if the write off group exists
+							write_off_groups
+								.get(index as usize)
+								.ok_or(Error::<T>::ErrInvalidWriteOffGroupIndex)?;
+							loan_data.admin_written_off = true;
+							Ok(index)
+						}
+						(admin_written_off, None) => {
+							// non-admin is trying to write off but admin already did. So error out
+							if admin_written_off {
+								return Err(Error::<T>::ErrLoanWrittenOffByAdmin.into());
+							}
+
+							let maturity_date = loan_data
+								.loan_type
+								.maturity_date()
+								.ok_or(Error::<T>::ErrLoanTypeInvalid)?;
+
+							// ensure loan's maturity date has passed
+							ensure!(now > maturity_date, Error::<T>::ErrLoanHealthy);
+
+							// not written off by admin, and non admin trying to write off, then
+							// fetch the best write group available for this loan
+							math::valid_write_off_group(
+								maturity_date,
+								now,
+								write_off_groups.clone(),
+							)
+							.ok_or(Error::<T>::ErrNoValidWriteOffGroup)
+						}
+					}?;
```

Comment:
Note that we are matching on the tuple `(loan_data.admin_written_off, override_write_off_index)` but you just case on the index.


```suggestion
					match override_write_off_index {
						// admin is trying to write off
						Some(index) => {
							// check if the write off group exists
							write_off_groups
								.get(index as usize)
								.ok_or(Error::<T>::ErrInvalidWriteOffGroupIndex)?;
							loan_data.admin_written_off = true;
							Ok(index)
						}
						
						None => {
							// non-admin is trying to write off but admin already did
							if loan_data.admin_written_off {
								return Err(Error::<T>::ErrLoanWrittenOffByAdmin.into());
							}

							let maturity_date = loan_data
								.loan_type
								.maturity_date()
								.ok_or(Error::<T>::ErrLoanTypeInvalid)?;

							// ensure loan's maturity date has passed
							ensure!(now > maturity_date, Error::<T>::ErrLoanHealthy);

							// not written off by admin, and non admin trying to write off, then
							// fetch the best write group available for this loan
							math::valid_write_off_group(
								maturity_date,
								now,
								write_off_groups.clone(),
							)
							.ok_or(Error::<T>::ErrNoValidWriteOffGroup)
						}
					}?;
```